packages/loopring_v3/contracts/impl/DelayedTransaction.sol

```diff
@@ -102,14 +105,6 @@ abstract contract DelayedTransaction is IDelayedTransaction, ReentrancyGuard
             transaction.value,
             transaction.data
         );
-
-        // Return the same data the original transaction would
```

Comment:
@Brecht, it seems we also have similiar code in Hebao's BaseWallet.sol:

```solidity
    /// @dev This default function can receive Ether or perform queries to modules
    ///      using bound methods.
    fallback()
        external
        payable
    {
        if (msg.data.length == 0 || msg.value > 0) {
            // Note: don't do anything here so send/transfer still works
            // -- the 2300 gas limit for send/transfer may case out-of-gas
            // issue after Istanbul.
            return;
        }

        address module = methodToModule[msg.sig];
        require(isAddressInSet(MODULE, module), "MODULE_UNAUTHORIZED");

        (bool success, bytes memory returnData) = module.call(msg.data);
        assembly {
            switch success
            case 0 { revert(add(returnData, 32), mload(returnData)) }
            default { return(add(returnData, 32), mload(returnData)) }
        }
    }
```

BTW, what's the difference between the old and the new implementation? It seems to me the old code will return some data in case of success, while the new code does not.

Reply:
The problem is that doing the return in assembly returns immediately (without solidity knowing), so the part of the code in `nonReentrant` after the function body is never executed. So the flag set to true for reentrance is never reset after the function ends, so all function calls that are done afterwards on `nonReentrant` functions will fail because the state is stuck in an invalid state where the contract thinks all calls are reentrant.

What the return does is to return whatever the function call returned even though the solidity function doesn't have a return value. So for this code in BaseWallet it's extremely important it is done because we really need the return value here to do what we want. We want to allow modules to immediately return data as if the wallet itself provided the function.

For the DelayedTransactions the return value is not important at all because it will just be used to update some state in a contract and the return value (in the rare case a function would even return some value) won't be used onchain anyway.

So instead of working around the problem by writing special cases for this I just removed that functionality to make it less error prone.

Reply:
Are you saying the following code in a *callee* contract will cause the *caller* contract to end up with `_guard ==1` and thus because unusable?
``` 
assembly {	
            switch success	
            case 0 { revert(add(returnData, 32), mload(returnData)) }	
            default { return(add(returnData, 32), mload(returnData)) }	
}
```

If so there is a big issue as the hacker can create a contract using the above assembly code, and put a lot of DeFI project into bad states...

Reply:
No, luckily not.  The *caller* and *callee* are separate states. The `return` will immediately return from the function call, not the complete transaction.

So the only way things can go wrong is inside the contract itself, not over multiple contracts. What happens is basically this:

```
foo()
   external
{
    require(reentrance == false);
    reentrance = true;

    // Early out in assembly not detected by solidity so `nonReentrant` doesn't 
    // work as expected. Using the normal return like we do here would actually 
    // work correctly with `nonReentrant` because solidity fully supports
    // this for modifiers. But solidity doesn't know the return happens
    // when doing it in assembly.
    return;

    reentrance = false;
}
```

The reason it does work as expected for the `revert` (which also exits the function immediately) is that a revert automatically reverts ALL state changes done in the function call anyway, so `reentrance` is reset this way. In this case the *caller* can continue on with the transaction if needed by catching the throw, but the *callee* is left in a valid state.